### PR TITLE
Combinatorial Fuzzing -- Fuzz with multiple mutations simultaneously

### PR DESCRIPTION
**Pull request by jtpereyda**
Added ability to fuzz combinations of multiple mutations. Finally! Also:

-  Added a Simple primitive that only fuzzes using the fuzz values you provide.
-  Added `examples/simple.py` to make it easy to demo the combinatorial functionality, e.g. with `python simple.py fuzz --record-passes 0 --text-dump --no-tui --target 127.0.0.1:2121 simple` (use `nc -vkl localhost 2121` to run a demo target)
- `Fuzzable.get_mutations()` now yields lists of Mutation objects, and each Mutation object can refer to exactly one mutation.
- MutationContext has a `mutations` attribute which is a map of qualified names to Mutation objects.
- `MutationContext` has the `message_path` property instead of `Mutation`.
- Added a couple CLI args for convenience
- Fixed a couple memory leaks in `fuzz_logger.py`
- `--test-case-name` now works for a single particular test case, making it easy to re-run specific cases
- Several web UI adaptations, but it could really use a redesign sometime
- Surprisingly few unit test changes needed.
- Note: Test case indexes will become less reliable when you start fuzzing with combinations of mutations. This could be fixed with a lot of effort, but might not yield sufficient value.
- Note: For the time being, multiple mutations are only applied to one message at a time.

Boofuzz now has a lot of the primitive functionality needed to incorporate code coverage; not sure when that will happen though.

Edit: Some details we'll probably want soon:

- [ ] The skip-after-N-failures functionality may not play well with multiple mutations right now.

### CODE REVIEW

**Review by SR4ven — CHANGES_REQUESTED** [review]
Nice work @jtpereyda!

Some things I noticed:

In the simple.py example we now fuzz all non-combinatorial cases for message1, then all for message1->message2, then we continue with the combinatorial cases for message1, and finally those for message1->message2.
Wouldn't be more straight forward to first fuzz all cases of a request before switching to the next one? Or is that the point you mentioned that's just too much work?

Also the progress bar in the webinterface seems to be broken, showing either Infinity or NaN:
image

Changelog entries are missing.

A simple request like...
```python
s_initialize("Test")
if s_block_start("test1"):
    s_delim("1")
    s_delim("2")
s_block_end("Test")
```
...now yields way more cases than expected.
From the log: `Info: Type: Delim. Default value: b''. Case 3960 of 88 overall.`

That leads to my next question, is the combinatorial fuzzing enabled for all primitives now? Because if so we would now generate a lot more cases than before. Even with just two delims from the example above we yield about 50 times more cases.
I see this problematic for large request definitions which would suddenly generate millions of cases which one simply cannot send to a target in an acceptable time.
Would it be possible to add a switch to session or perhaps fuzzable to enable/disable combinatorial fuzzing?

File: boofuzz/primitives/simple.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+from ..fuzzable import Fuzzable
+
+
+class Simple(Fuzzable):
```
[the remark this answers is not in the document]
Comment (jtpereyda) [reply boofuzz/primitives/simple.py:4]: Good point on both -- might as well maintain the old approach I suppose...

**Comment by jtpereyda** [comment]
> Wouldn't be more straight forward to first fuzz all cases of a request before switching to the next one? 

For a typical real world protocol message, this could take an effectively infinite amount of time. Say you have 10 elements 1k mutations each. Fuzzing for x2 mutations yields around 1k*1k*(10 choose 2 == 45) == 45M mutations. Fuzzing for x3 yields 1k*1k*(10 choose 3) == 120M. Etc. So the current strategy is to go sort of "breadth first" by fuzzing all of the protocol for one mutation, then trying to fuzz all the protocol at 2 mutations, etc.

Now, even this is semi-naive, as a true breadth-first approach would jump around elements and around the protocol more randomly. But it's a start.

Rather than investing in smarter random strategies, a better next step may be to incorporate code coverage feedback.

> ...now yields way more cases than expected.
> From the log: Info: Type: Delim. Default value: b''. Case 3960 of 88 overall.

Yes, the estimate still shows the number of 1-mutations. But the test will keep on going well past that.

> Also the progress bar in the webinterface seems to be broken, showing either Infinity or NaN:

Hmm, I'll look at it; for me it's showing the 1-mutation estimate, and the number on the top-right keeps growing right past 100%.

> That leads to my next question, is the combinatorial fuzzing enabled for all primitives now? Because if so we would now generate a lot more cases than before. Even with just two delims from the example above we yield about 50 times more cases.
> I see this problematic for large request definitions which would suddenly generate millions of cases which one simply cannot send to a target in an acceptable time.
> Would it be possible to add a switch to session or perhaps fuzzable to enable/disable combinatorial fuzzing?

We could add a switch. This brings boofuzz closer in behavior to mainstream fuzzers like AFL that fuzz indefinitely by default. It's not truly infinite as you can run out of cases, but in some (most?) practical situations you never would run out.

**Comment by SR4ven** [comment]
Ok got it, first we fuzz the protocol like before and then we try to generate as many cases as possible by combining them.
I was just confused that it kept going after the expected number of test cases. And I assume others will run into the same issue so we should try to make the new behaviour more clear.
Printing "case 3000 of 80" just doesn't feel right but I can't come up with a better way. Maybe just skip the total number once we reach the expected test case number.
The whole progress bars and number of mutations part becomes a little misleading with this feature. Maybe we can calculate/estimate the number of combinatorial cases? And then display two stages/progress bars for non combinatorial and combinatorial?

>Rather than investing in smarter random strategies, a better next step may be to incorporate code coverage feedback.

Code coverage feedback would be lovely indeed.

>We could add a switch though. Maybe --combinatorial/--no-combinatorial or something.

Sounds good. When running boofuzz in CI it's definitely useful to have a switch to reduce the test case count.

**Comment by jtpereyda** [comment]
Review suggestions applied. The log no longer contains "Case N of M" if --combinatorial is used. Likewise, the web page will no longer display progress bars and giant percentages.

**Review by SR4ven — APPROVED** [review]
LGTM

**Pull request merged**